### PR TITLE
Editor: prevent post.type from being null when transitioning to saved.

### LINKED ISSUE #7023: CPT / Editor: Add New Project Type Module disappears when project auto-saves
#### Steps to reproduce
1. Starting at URL: https://wordpress.com/edit/jetpack-portfolio/$site
2. Add a project title
3. (Quickly) open up the Add New Project Type module
4. With module open, watch the project automatically save and the module disappears

#### What I expected

I expected the module to remain in view and allow edit, the way it does with regular Posts

#### What happened instead

The module disappears without saving changes

#### Screenshot / Video

https://cloudup.com/cGV_7XI8rAI

### PR DESCRIPTION
**Pull request by timmyc**
Fixes #7023 - when a post is transitioning from an un-saved draft to saved ( either via an autoSave or manually clicking the save button ), there were a few render calls being made to `<EditorDrawerTaxonomies />` where the `post.type` was undefined, and therefore no `taxonomies` were being passed to the render call.

This resulted in the component rendering empty, and then re-rendering with all the correct `taxonomies` for the post type.

For the fix I have changed the order of the dispatches in `onSaveSuccess`, and did a small bit of lint fixing in that function as well.

__To Test__
- Open up a new portfolio item in the editor
- Expand Project Types accordion in the sidebar
- add a title, and some content and click save
- verify the Project Types accordion is still open

@aduth can you think of any flows where changing the order of these actions might cause issues?

Test live: https://calypso.live/?branch=fix/editor/7023

### CODE REVIEW

**Comment by aduth** [comment]
>For the fix I have changed the order of the dispatches in `onSaveSuccess`

What's the reason that this fixes the issue? Is it that when we reset after receiving the saved post, attempts to retrieve the post value will correctly fall back to the now-updated post state?

File: client/post-editor/post-editor.jsx [hunk 1]
```diff
@@ -696,7 +691,12 @@ const PostEditor = React.createClass( {
 		// Receive updated post into state
 		this.props.receivePost( post );
 
-		nextState = {
+		// Reset previous edits, preserving type
+		this.props.resetPostEdits( this.props.siteId );
+		this.props.resetPostEdits( post.site_ID, post.ID );
+		this.props.editPost( { type: this.props.type }, post.site_ID, post.ID );
```
Comment (aduth) [inline client/post-editor/post-editor.jsx]: I wonder if we need this after all?

**Comment by timmyc** [comment]
> Is it that when we reset after receiving the saved post, attempts to retrieve the post value will correctly fall back to the now-updated post state?

Correct, since `EditorDrawerTaxonomies` uses `getEditorPostId` along with `getEditedPostValue` to obtain the `type` of post, the prior order was resulting in `undefined` values.  By calling `setEditorPostId` and  `receivePost` first, the postId and values ( including `type` ) are present in the connect call.

**Comment by aduth** [comment]
The change seems sensible, and definitely explains some of the strange behavior that occurs on save. In particular, it also seems to resolve the issue of the URL flickering briefly to `/post` for custom post types upon a save.

I even tested with [removing the `editPost` for `type`](https://github.com/Automattic/wp-calypso/pull/7050#discussion_r72342760) and it still seems to work as expected; which makes sense, since attempts to retrieve `type` via `getEditedPostValue` should fall back to the saved post value. Wonder if this will get us one step closer to removing [this hard-coded condition for `type` in dirty checking](https://github.com/Automattic/wp-calypso/blob/f1e9410f91a64be62f9dd14bb9c8061f8322a7e2/client/state/posts/selectors.js#L348-L350) (cc @nylen , related: #6866)

**Comment by timmyc** [comment]
Removed the `editPost` call in eafb3fa and indeed things still work well.  I backed out the change in #6866, but the dirty state still exists on a new page - so a bit more work is needed there.

**Comment by aduth** [comment]
>I backed out the change in #6866, but the dirty state still exists on a new page - so a bit more work is needed there.

Yeah, I guess it's reflecting the `type` edit [dispatched in the controller for a new page](https://github.com/Automattic/wp-calypso/blob/d2d0a3176fbe22269427859a6f7dfecb5fee4367/client/post-editor/controller.js#L118). I don't see an obvious way that we can get rid of that currently, but in any case, it's not really relevant to the changes here so I'm fine to leave for a later PR.

This is looking good and working well for me 👍

**Pull request merged**